### PR TITLE
add linear lr warmup and lr decay scheduler

### PR DESCRIPTION
**Pull request by lessw2020**
this PR adds a linear lr scheduler and includes some automation based on current best practices:
a - takes user lr provided in args as lr_max, and computes final min_lr for the decay schedule based on lr / 10, per chinchilla paper.  (i.e. total decay will be one order of magnitude).
b - computes an automated linear warmup schedule of 10% total iters as warmup, with min warmup of 2 steps. 
c - computes a linear decay schedule after warmup, declining from lr_max to lr_min over the end of warmup to end of training.  (per Aarons latest paper, linear is preferred schedule). 
d - I updated learning rate to 8e-4, in order to provide more visible per iter results to the user assuming debugModel. 

LR scheduling produces much improved loss curve:

Screenshot 2024-01-28 at 6 39 34 PM

I added two log prints - the warmup schedule as one line, and then a step and current lr at each iter.  
Both could be disabled if too much info.

### CODE REVIEW

File: torchtrain/lr_scheduling.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        # enforce min of 2 steps for warmup
+        self.warmup_steps = max(int(args.steps * self.lr_warmup_pct), 2)
+
+        rank0_log(
+            f"LR Warmup Schedule: {self.lr_min} -> {self.lr_max} with {self.warmup_steps} warmup steps"
+        )
+        self.decay_steps = args.steps - self.warmup_steps
+        self.curr_lr = 0
+
+    def set_lr(self, optimizer, step):
```
Comment (wanchaol) [inline torchtrain/lr_scheduling.py]: I think it's better to use the pytorch optim lr scheduler API instead

Comment (lessw2020) [reply torchtrain/lr_scheduling.py]: I actually initially wrote this using a chained scheduler to combine warmup and then linear decay - however the torch functions don't have the concept of a min lr.   In otherwords, you would decay to 0.0 learning rate and can't do the chinchilla recommended approach of decaying from peak to min by factor of 10.

Comment (lessw2020) [reply torchtrain/lr_scheduling.py]: Thus wrote one that we have direct control over.

File: torchtrain/lr_scheduling.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+
+from torch.optim.lr_scheduler import LambdaLR
+
+# global states for scheduling
+# these are needed as LambdaLR does not support argument passing
+_warmup_steps = 2
```
Comment (wanchaol) [inline torchtrain/lr_scheduling.py:8]: nit: we can make these two arguments still as function arguments below, but we use:

`functools.partial` to pack these two arguments when adding to the optimizer. 

global config works for now. This can be done as a follow up

File: torchtrain/lr_scheduling.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+    else:
+        # linear decay
+        normalized_step = _decay_steps - (current_step - _warmup_steps)
+        curr_adjustment = 1 - (_decay_steps - normalized_step) / _decay_steps
+
+    return curr_adjustment
+
+
+def get_full_lr_scheduler(optimizer, args):
```
Comment (wanchaol) [inline torchtrain/lr_scheduling.py]: what's "full_lr_scheduler" mean here? can we just call this as "get_lr_scheduler"?

File: train.py [hunk 4]
```diff
@@ -143,7 +147,10 @@ def main(args):
     parser.add_argument(
         "--optimizer", type=str, default="AdamW", help="optimizer to use"
     )
-    parser.add_argument("--lr", type=float, default=2e-5, help="learning rate to use")
+    parser.add_argument("--lr", type=float, default=8e-4, help="learning rate to use")
+    parser.add_argument(
+        "--warmup_pct", type=float, default=0.10, help="pct training to use for warmup"
```
Comment (wanchaol) [inline train.py]: what is "pct" mean here? can we elaborate more in the help text?

File: torchtrain/lr_scheduling.py (hunk 2 above)
Comment (lessw2020) [reply torchtrain/lr_scheduling.py:8]: created an issue to track this:  https://github.com/pytorch-labs/torchtrain/issues/29

File: torchtrain/lr_scheduling.py (hunk 3 above)
Comment (lessw2020) [reply torchtrain/lr_scheduling.py]: I had full in there to express that it is both a warmup scheduler and decay scheduler...but yes simpler to just call it get_lr_scheduler.
updated!

File: train.py (hunk 4 above)
Comment (lessw2020) [reply train.py]: updated tde help text - "help="percentage of total training steps to use for warmup"
basically how much of the total training time should be run under the warmup lr schedule vs decay schedule.  People often use fixed steps for warmup, but 1500 steps warmup is pointless if only running 200 iters...thus keeping it as a pct makes it dynamic to any training schedule.
